Codegen: Registration now only includes the functions used

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69840
* #68462
* #68693
* #68692
* #68691
* #68697
* #68690
* __->__ #68689
* #68714
* #68688
* #68687
* #68247

Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

peterbell10 committed:
Codegen: Registration now only includes the functions used

Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/SortingUtils.h:
```diff
@@ -4,6 +4,8 @@
 #include <ATen/native/Resize.h>
 #include <c10/util/irange.h>
 
+#include <ATen/ops/empty.h>
```
Do you have an automatic way to populate these? Do you think we want to (can?) have one so that we can keep these minimal over time?

bdhirsh replied:
@albanD I remember Ed mentioning some problems with his ufunc codegen PR around internal buck targets not being able to handle dynamically codegen'd headers. Did that not turn out to be a problem here with the per-op headers?

albanD replied:
I just imported this stack. Will see how it goes. Hopefully not...
But potentially, we can gate it on cmake version / internal and the internal incremental build will just be not as good.

peterbell10 replied:
The headers required special handling to work with `buck`/`bazel`. See the build system changes here:
https://github.com/pytorch/pytorch/pull/68247/files#diff-ff6cf959d408c4356a73d786776adac2a9a39a90ed1689f392e8ec23984e0764

`genrule` ("general rule") requires the list of output files to be statically known, but with  `rule` you can create your own rules that are more flexible. Here the only difference is I specify an entire directory as an output. This has shown to work this with `bazel` locally and in CI, hopefully `buck` doesn't diverge too much here.

peterbell10 replied:
>Do you have an automatic way to populate these? Do you think we want to (can?) have one so that we can keep these minimal over time?

That would be really useful. I know there is [include-what-you-use](https://include-what-you-use.org/) which does this for headers generally. Not sure if you're able to limit it to run on specific kinds of includes.

albanD commented on aten/src/ATen/templates/RegisterDispatchKey.cpp:
```diff
@@ -5,12 +5,19 @@
+    defined(TORCH_HIP_BUILD_MAIN_LIB)     || \
+    defined(TORCH_CUDA_CU_BUILD_MAIN_LIB) || \
+    defined(TORCH_CUDA_CPP_BUILD_MAIN_LIB)
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
```
Where does this not work? should we leave a comment here for the next person who's going to try and fix it?
(sorry if this is in a PR up the stack, I'm reviewing on the way up)

peterbell10 replied:
This special case only excludes external projects such as XLA which re-use some of pytorch's codegen, including `RegisterDispatchKey.cpp`.

albanD commented on tools/codegen/dest/__init__.py:
```diff
@@ -1,3 +1,6 @@
+from .register_dispatch_key import (
+    RegisterDispatchKey as RegisterDispatchKey,
+    gen_registration_helpers as gen_registration_helpers,
+    gen_registration_headers as gen_registration_headers,
```
Why the `as` here?

peterbell10 replied:
`mypy` needs it for some reason. It doesn't make sense to me either.

albanD reviewed (commented):
Sounds good!

cc @bdhirsh do you want to take a look as well?

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Rebase to fix merge conflicts on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596273).

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Fix mypy issues on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Commit to re-run CI on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Registration now only includes the functions used"


Currently Register{DispatchKey}.cpp includes all of
`NativeFunctions.h`, so any operator signature change requires all
backend registration to be recompiled. However, most backends only
have registrations for a small fraction of operators so it makes sense
to only include the specific functions required.

Differential Revision: [D32596273](https://our.internmc.facebook.com/intern/diff/D32596273)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596273).

albanD commented on tools/codegen/gen.py:
```diff
@@ -171,14 +171,40 @@ def static_dispatch_keys(backend: Optional[BackendIndex]) -> List[DispatchKey]:
+        # TODO: for ops with structured_delegate it should check the dispatch table of
+        # the out variant instead. For now, these structured ops all have CPU/CUDA kernels
+        # so we always dispatch to the `backend`, but this could be wrong when we
+        # migrate math/default_backend ops to use structured delegate.
```
cc @bdhirsh on this one.

peterbell10 replied:
In case it isn't clear, this `TODO` comes from the existing function `static_dispatch` below. I've just factored it out into a separate function so I can use it to include the appropriate static dispatch header.

albanD replied:
Yes this is just a move. But Brian is working on this part of the codegen. So wanted to flag this for him in case he never saw this comment before.

bdhirsh replied:
Sounds good. Mobile is working on some changes to static dispatch soon so we might have some merge conflicts. I'll let them know.

ezyang commented on tools/codegen/gen.py:
```diff
@@ -1275,16 +1300,46 @@ def gen_source_files(
+                        headers.append(
+                            f"#include <ATen/ops/{fn.root_name}_{dispatch_namespace}_dispatch.h>")
+
+                return sorted(set(headers))
```
Is there any particular reason you coalesced multiple overloads together into a single header by root_name? Besides aesthetics, it seems strictly worse, since there isn't any intrinsic relationship between various overloads and they can be profitably split into separate headers. Would probably simplify the grouping logic elsewhere as well.

peterbell10 replied:
This might be good to discuss in pytorch/rfcs#39.  The main reason for me is I don't think a programmer should need to know the name of the overload they're calling, which you would in order to include it. Of course this is a trade off for more files potentially being rebuilt when an overload is changed but I am doubtful this is a significant issue e.g. most overloads are implemented in the same file so would get rebuilt together anyway.

I suppose there could be a hierarchy of headers where we start with one per overload and then group headers by `root_name` then into the top level header like  `Functions.h`. However, that would be a massive jump in file count which might lead to problems of their own.

ezyang replied:
Well, I discovered one good reason to coalesce: it makes the logic here simpler for grouped native functions (patch coming soon)
